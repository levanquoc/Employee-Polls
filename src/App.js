import React from 'react';
import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
import Login from './components/Login';
import Home from './components/Home';
import Leaderboard from './components/Leaderboard';
import Menu from './components/Menu';
import { useSelector } from 'react-redux';
import Add from './components/Add';
import Question from './components/Question';
import PrivateRoute from './components/PrivateRoute';
import Private2Route from './components/Private2Route';
import NotFound from './components/404';

function App() {
  const isAuthenticated = useSelector((state) => state.auth.isAuthenticated);
  return (
    <div>
      <Router>
        {isAuthenticated && <Menu/>}
        <Routes>
          <Route path="/login" element={<Private2Route><Login /></Private2Route>} />
          <Route path="/home" element={<PrivateRoute><Home /></PrivateRoute>} />
          <Route path="/leaderboard" element={<PrivateRoute><Leaderboard /></PrivateRoute>} />
          <Route path="/add" element={<PrivateRoute><Add /></PrivateRoute>} />
          <Route path="/questions/:questionId" element={<PrivateRoute><Question /></PrivateRoute>} />
          <Route path="*" element={<PrivateRoute><NotFound /></PrivateRoute>} />
        </Routes>
      </Router>
    </div>
  );
}
 
export default App;