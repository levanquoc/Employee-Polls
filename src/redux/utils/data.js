let users = {
    sarahedo: {
      id: 'sarahedo',
      password:'password123',
      name: 'Sarah Edo',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS1nYUKHA8kkUlIr9DG9woo_y0QtY9m9SGylQ&s',
      answers: {
        "8xf0y6ziyjabvozdd253nd": 'optionOne',
        "6ni6ok3ym7mf1p33lnez": 'optionOne',
        "am8ehyc8byjqgar0jgpub9": 'optionTwo',
        "loxhs1bqm25b708cmbf3g": 'optionTwo'
      },
      questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
    },
    tylermcginnis: {
      id: 'tylermcginnis',
      password:'abc321',
      name: 'Tyler McGinnis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRV1Mly7C6D_WWpPXTAO4dF52D9Wd9FKuC9zw&s',
      answers: {
        "vthrdm985a262al8qx3do": 'optionOne',
        "xj352vofupe1dqz9emx13r": 'optionTwo',
      },
      questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
    },
    mtsamis: {
      id: 'mtsamis',
      password:'xyz123',
      name: 'Mike Tsamis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcScuQGyYbgV9HFyiunO9mF6_lnB6MYwcx6t3w&s',
      answers: {
        "xj352vofupe1dqz9emx13r": 'optionOne',
        "vthrdm985a262al8qx3do": 'optionTwo',
        "6ni6ok3ym7mf1p33lnez": 'optionOne'
      },
      questions: ['6ni6ok3ym7mf1p33lnez', 'xj352vofupe1dqz9emx13r'],
    },
    zoshikanlu: {
      id: 'zoshikanlu',
      password:'pass246',
      name: 'Zenobia Oshikanlu',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQGLRAlpUwgnA01Ksosn99mzvGi1dZEeLS0Mw&s',
      answers: {
        "xj352vofupe1dqz9emx13r": 'optionOne',
      },
      questions: [],
    }
  }
  
  let questions = {
    "8xf0y6ziyjabvozdd253nd": {
      id: '8xf0y6ziyjabvozdd253nd',
      author: 'sarahedo',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS1nYUKHA8kkUlIr9DG9woo_y0QtY9m9SGylQ&s',
      timestamp: 1467166872634,
      optionOne: {
        votes: ['sarahedo'],
        text: 'Build our new application with Javascript',
      },
      optionTwo: {
        votes: [],
        text: 'Build our new application with Typescript'
      }
    },
    "6ni6ok3ym7mf1p33lnez": {
      id: '6ni6ok3ym7mf1p33lnez',
      author: 'mtsamis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcScuQGyYbgV9HFyiunO9mF6_lnB6MYwcx6t3w&s',
      timestamp: 1468479767190,
      optionOne: {
        votes: [],
        text: 'hire more frontend developers',
      },
      optionTwo: {
        votes: ['mtsamis', 'sarahedo'],
        text: 'hire more backend developers'
      }
    },
    "am8ehyc8byjqgar0jgpub9": {
      id: 'am8ehyc8byjqgar0jgpub9',
      author: 'sarahedo',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS1nYUKHA8kkUlIr9DG9woo_y0QtY9m9SGylQ&s',
      timestamp: 1488579767190,
      optionOne: {
        votes: [],
        text: 'conduct a release retrospective 1 week after a release',
      },
      optionTwo: {
        votes: ['sarahedo'],
        text: 'conduct release retrospectives quarterly'
      }
    },
    "loxhs1bqm25b708cmbf3g": {
      id: 'loxhs1bqm25b708cmbf3g',
      author: 'tylermcginnis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRV1Mly7C6D_WWpPXTAO4dF52D9Wd9FKuC9zw&s',
      timestamp: 1482579767190,
      optionOne: {
        votes: [],
        text: 'have code reviews conducted by peers',
      },
      optionTwo: {
        votes: ['sarahedo'],
        text: 'have code reviews conducted by managers'
      }
    },
    "vthrdm985a262al8qx3do": {
      id: 'vthrdm985a262al8qx3do',
      author: 'tylermcginnis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRV1Mly7C6D_WWpPXTAO4dF52D9Wd9FKuC9zw&s',
      timestamp: 1489579767190,
      optionOne: {
        votes: ['tylermcginnis'],
        text: 'take a course on ReactJS',
      },
      optionTwo: {
        votes: ['mtsamis'],
        text: 'take a course on unit testing with Jest'
      }
    },
    "xj352vofupe1dqz9emx13r": {
      id: 'xj352vofupe1dqz9emx13r',
      author: 'mtsamis',
      avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcScuQGyYbgV9HFyiunO9mF6_lnB6MYwcx6t3w&s',
      timestamp: 1493579767190,
      optionOne: {
        votes: ['mtsamis', 'zoshikanlu'],
        text: 'deploy to production once every two weeks',
      },
      optionTwo: {
        votes: ['tylermcginnis'],
        text: 'deploy to production once every month'
      }
    },
  }
  
  function generateUID () {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15)
  }
  
  export function _getUsers () {
    return new Promise((resolve) => {
      setTimeout(() => resolve({...users}), 1000)
    })
  }
  
  export function _getQuestions () {
    return new Promise((resolve) => {
      setTimeout(() => resolve({...questions}), 1000)
    })
  }
  
  function formatQuestion ({ optionOneText, optionTwoText, author, avatarURL }) {
    return {
      id: generateUID(),
      timestamp: Date.now(),
      author,
      avatarURL,
      optionOne: {
        votes: [],
        text: optionOneText,
      },
      optionTwo: {
        votes: [],
        text: optionTwoText,
      }
    }
  }
  
  export function _saveQuestion (question) {
    return new Promise((resolve, reject) => {
      if (!question.optionOneText || !question.optionTwoText || !question.author) {
        reject("Please provide optionOneText, optionTwoText, and author");
        return ;
      }
  
      const formattedQuestion = formatQuestion(question)
      const authorId = question.author
      
      setTimeout(() => {
        questions = {
          ...questions,
          [formattedQuestion.id]: formattedQuestion
        }

        users = {
          ...users,
          [authorId]: {
            ...users[authorId],
            questions: users[authorId].questions.concat([formattedQuestion.id])
          }
        }
  
        resolve(formattedQuestion)
      }, 1000)
    })
  }
  
  export function _saveQuestionAnswer ({ authedUser, qid, answer }) {
    return new Promise((resolve, reject) => {
      if (!authedUser || !qid || !answer) {
        reject("Please provide authedUser, qid, and answer");
        return;
      }
  
      setTimeout(() => {
        users = {
          ...users,
          [authedUser]: {
            ...users[authedUser],
            answers: {
              ...users[authedUser].answers,
              [qid]: answer
            }
          }
        }
  
        questions = {
          ...questions,
          [qid]: {
            ...questions[qid],
            [answer]: {
              ...questions[qid][answer],
              votes: questions[qid][answer].votes.concat([authedUser])
            }
          }
        }
  
        resolve(true)
      }, 500)
    })
  }
  