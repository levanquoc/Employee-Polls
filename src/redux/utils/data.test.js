/**
 * @jest-environment jsdom
 */
import * as React from 'react';
import renderer from 'react-test-renderer';
import App from '../../App';
import { _saveQuestion,_saveQuestionAnswer  } from './data';
import { Provider } from 'react-redux';
import store from '../store';

describe('Testing with function _saveQuestion', () => {
    it('should return the saved question with all expected fields when correctly formatted data is passed', async () => {
        const question = {
          optionOneText: 'Option One',
          optionTwoText: 'Option Two',
          author: 'sarahedo',
          avatarURL: 'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcS1nYUKHA8kkUlIr9DG9woo_y0QtY9m9SGylQ&s'
        };
    
        const savedQuestion = await _saveQuestion(question);
        expect(savedQuestion).toHaveProperty('id');
        expect(savedQuestion).toHaveProperty('timestamp');
        expect(savedQuestion).toHaveProperty('author', question.author);
        expect(savedQuestion).toHaveProperty('avatarURL', question.avatarURL);
        expect(savedQuestion).toHaveProperty('optionOne');
        expect(savedQuestion.optionOne).toHaveProperty('text', question.optionOneText);
        expect(savedQuestion.optionOne).toHaveProperty('votes');
        expect(savedQuestion).toHaveProperty('optionTwo');
        expect(savedQuestion.optionTwo).toHaveProperty('text', question.optionTwoText);
        expect(savedQuestion.optionTwo).toHaveProperty('votes');
    });

    it('should return an error if incorrect data is passed:optionOneText', async () => {
        const question = {
            optionOneText: 'Option One',
            // Thiếu optionTwoText và author
        };
        await expect(_saveQuestion(question)).rejects.toEqual('Please provide optionOneText, optionTwoText, and author');
    });
    it('should return an error if incorrect data is passed:optionTwoText', async () => {
        const question = {
            optionTwoText: 'Option Two',
            // Thiếu optionTwoText và author
        };
        await expect(_saveQuestion(question)).rejects.toEqual('Please provide optionOneText, optionTwoText, and author');
    });
});

describe('Testing with function _saveQuestionAnswer', () => {
    it('should return true when correctly formatted data is passed', async () => {
        const authedUser = 'sarahedo';
        const qid = '8xf0y6ziyjabvozdd253nd';
        const answer = 'optionOne';
    
        const result = await _saveQuestionAnswer({ authedUser, qid, answer });
        expect(result).toBe(true);
    });

    it('should return an error if incorrect data is passed : authedUser', async () => {
        const authedUser = '';
        const qid = '123123'; // qid bị bỏ trống
        const answer = 'optionOne';

        await expect(_saveQuestionAnswer({ authedUser, qid, answer })).rejects.toEqual("Please provide authedUser, qid, and answer");
    });
    it('should return an error if incorrect data is passed : qid', async () => {
        const authedUser = 'sarahedo';
        const qid = ''; // qid bị bỏ trống
        const answer = 'optionOne';

        await expect(_saveQuestionAnswer({ authedUser, qid, answer })).rejects.toEqual("Please provide authedUser, qid, and answer");
    });
    it('should return an error if incorrect data is passed : answer', async () => {
        const authedUser = 'sarahedo';
        const qid = '12321'; // qid bị bỏ trống
        const answer = '';

        await expect(_saveQuestionAnswer({ authedUser, qid, answer })).rejects.toEqual("Please provide authedUser, qid, and answer");
    });
});

describe('Testing with toMatchSnapshot', () => {
    it('matches the snapshot when a name is passed', () => {
        const component = renderer.create(
            <Provider store={store}>
                     <App />
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});