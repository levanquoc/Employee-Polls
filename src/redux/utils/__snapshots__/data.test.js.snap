// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing with toMatchSnapshot matches the snapshot when a name is passed 1`] = `<div />`;
